### ISSUE TITLE
[Feature Request] New `BeartypeConf(hint_mappings)` option. @felixchenier deserves it!

### ISSUE THREAD

**Issue by felixchenier** [beartype/beartype#302]
Hello hello

Here is me, coming with bad numpy stuff.

This code creates a numpy array of integers, then indexes one of these integers to be printed by a function that accepts integers as an input. End of the story.

```python
import numpy as np
from beartype import beartype, BeartypeConf

typecheck = beartype(conf=BeartypeConf(is_pep484_tower=True))


@typecheck
# Functions
def print_integer(var: int):
    print(var)


# Stuff
if __name__ == "__main__":
    table = np.array([1, 1])
    i = table[0]
    print(type(i))
    
    print_integer(i)
```

But to my grandiose sadness, numpy defines its own integer class and everything falls apart.

```python
<class 'numpy.int64'>
Traceback (most recent call last):

  File ~/opt/miniconda3/envs/mosa/lib/python3.10/site-packages/spyder_kernels/py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File ~/Documents/git/kineticstoolkit/kineticstoolkit/untitled0.py:19
    print_integer(i)

  File <@beartype(__main__.print_integer) at 0x15e519990>:21 in print_integer

BeartypeCallHintParamViolation: Function __main__.print_integer() parameter var=1 violates type hint <class 'int'> under non-default configuration BeartypeConf(claw_is_pep526=True, is_color=None, is_debug=False, is_pep484_tower=True, strategy=<BeartypeStrategy.O1: 2>, warning_cls_on_decorator_exception=None), as <protocol "numpy.int64"> 1 not instance of int.
```

So beartype does the correct thing, but the correct thing is not correct. To whom should I address my prayers?

**Comment by leycec** [comment]
@posita: I politely summon thee to solve this desperate plea from a very nice man who has also submitted a [very nice PR](https://github.com/beartype/beartype/pull/301). In the hopes that this man will continue to submit very nice PRs, I wish to appease this man. I probably shouldn't have written exactly what I was thinking. But that's what happened anyway.

Thankfully, I *think* I already know what *probably* needs to be done here. Here's my uneducated **tl;dr**. To transparently support NumPy scalars under runtime type-checkers like @beartype, please do:

* **Add @posita's [wonderful open-source `numerary` package as an a mandatory dependency of your package](https://github.com/beartype/numerary).** As the URL suggests, `numerary` is an official @beartype project under the @beartype umbrella. This is why we @beartype.
* Do what I do and may @posita have mercy on your codebase:

```python
from numerary import IntegralLike  # <-- replace all "int" type hints with this for great glory

import numpy as np
from beartype import beartype, BeartypeConf

typecheck = beartype(conf=BeartypeConf(is_pep484_tower=True))


@typecheck
# Functions
def print_integer(var: IntegralLike):  # <-- "IntegralLike" intensifies
    print(var)
```

That's it. That's `numerary`. Avoid the builtin `int` type, which is dumb. Instead, prefer the @beartype-friendly `numerary.IntegralLike` type, which is smart. @posita made it! So you just know it works.

Praise be to @posita. He both giveth and taketh in equal measure.

![](https://media.tenor.com/Zu0ClSnqSXEAAAAC/blessing-from-the-lord-god-be-praised.gif)

**Comment by felixchenier** [comment]
Hi @leycec 

Thanks for this answer. Unfortunately I think I'm running in circle (#215). And honestly I'm increasingly doubtful in the very idea of real-time type checking, at least until Python improves this typing mess dramatically. `</inadvertently_slam_potential_friend_in_face>`
`</regret>`

I'm happy that you appreciate my PR, but really my current position on the circle I'm running on, is that I'm probably ready to quit this boat again, at least temporarily. If I continue running my circle, I will change my mind. But probably to change it again.

Here is my psychic state of mind, highly philosophical but explained as an engineer's vomit spill: https://github.com/kineticstoolkit/kineticstoolkit/issues/198#issuecomment-1801993900. I may be back, but for now I think I will consider the idea in this post more thoroughly (typecasting vs typechecking).

**Comment by felixchenier** [comment]
@leycec 

> If I continue running my circle, I will change my mind.

Already there. I run very fast.

And I must say your comment to my PR yesterday made me feel a bit cheap about abandoning the ship again, when I'm so close to success and when we still have one believer (you) who dedicates probably most of his life to type checking (I hope not!).

There's no urgency for me to change anything to my code at this point, so let's find the best solution. I have two questions regarding your proposition https://github.com/beartype/beartype/issues/302#issuecomment-1801110907

## 1. Do we really need numerary for this?

It seems that using numbers.Integral works as is:

```
isinstance(1, numbers.Integral)
Out[39]: True

isinstance(1.0, numbers.Integral)
Out[40]: False

isinstance(np.array([1])[0], numbers.Integral)
Out[41]: True

isinstance(np.array([1.0])[0], numbers.Integral)
Out[42]: False

isinstance(sympy.Integer(1), numbers.Integral)
Out[52]: True

isinstance(sympy.Float(1), numbers.Integral)
Out[53]: False
```

## 2. Could I avoid modifying my type hints?

I understand that I could do a "best" workaround by defining my new custom type `Int` as being `numbers.Integral` (or if I'm wrong with my point 1, as being `numerary.IntegralLike`). But then I create three new smaller problems:

- My type hints are not `int` anymore, they are `Int`. It may seem frivolous, but the package I distribute aims to train biomechanicians to programming, and I want to teach them best practices (you wouldn't believe what a novice scientist who is also novice at programming can write). If they see my functions in the documentation that use type hint `Int`, they will also use `Int` and wonder why it complains.
- Using a custom-made type hint `Int` implicitly means that we made some modification to the base type `int`, or that it still means an integer but has nothing to do with type `int`. This is not the case, and therefore we just introduced confusion.
- In addition to typecheck my type hints, I use sphinx to generate the documentation. It has the unfortunate tendency to expand type hints, so that type `Int` in my docstrings could end-up being `numerary.IntegralLike` in the website documentation, which is still more confusing. There are ways to avoid this, but I don't like put a plaster on top of another plaster.

Which leads me to what seems like a new feature request.

Beartype can already deal with type aliases. But we can't and shouldn't overload builtin types (everything would explose and we usually want to avoid explosions). But could we have a way to tell Beartype: hey, bear! When you read `int`, would you mind saying `numeral.Integral` instead?

Like: 
```
BeartypeConf(type_mappings: {int: numeral.Integral})
```

Is it cheating? Yes.

Is it really cheating? Maybe not, since it's still explicit.

Is is cheating then? No.

It's just an idea, but I think it solves many problems at once:

- We keep simple type hints for more precise user's documentation.
- Simple type hints do not make sphinx angry.
- Type checking works.

**Comment by leycec** [comment]
**Excellent idea!** Apologies for the significant delay, too. I've been putting out fires across the @beartype codebase like an overworked salaryman about to keel over from *[karoshi](https://en.wikipedia.org/wiki/Karoshi)*.

Yes, let's absolutely do this ingenious `type_mappings` option you have now invented for everyone. To expand the idea, let's generalize this to map arbitrary *type hints* rather than merely *types.* I'm now envisioning something vaguely similar to this:

```python
BeartypeConf(hint_mappings={
    int: numbers.Integral,
    list[str]: Sequence[str],  # <-- look, just do it @beartype!
})
```

I also know actually where and how to implement support for this in the @beartype codebase. If you're so inclined, <sup>*please be so inclined, bonn homme bro*</sup> see also:

* The `_reduce_hint_cached()` function in the `beartype._check.convert.convreduce` submodule: e.g.,

```python
@callable_cached
def _reduce_hint_cached(
    hint: Any,
    conf: BeartypeConf,
    exception_prefix: str,
) -> object:
    '''
    ...
    '''

    # If this configuration maps this hint to another hint, do so now.
    # This one-liner looks ridiculous, but actually works. More
    # importantly, this is the fastest way to accomplish this. Flex!
    hint = conf.hint_mappings.get(hint, hint)

    # Sign uniquely identifying this hint if this hint is identifiable *OR*
    # "None" otherwise.
    hint_sign = get_hint_pep_sign_or_none(hint)
```

Ridiculous one-liner is ridiculous. :flex: :bear: 

Thanks for the conclusive heads-up about `numbers.Integral`, too. NumPy once failed to register its scalars with abstract base classes (ABCs) in the standard `numbers` module. Thus, `numerary`. But you have just proved that NumPy now does that. My excitement is palpable and my palms are sweaty. :sweat_smile: 

Let's rename this issue to gently prod one of us to actually implement this exciting API someday.

**Comment by felixchenier** [comment]
@leycec This rocks! We will make beartype do great things, I know it.

Thanks for telling me where to put the code, I tried to find myself but I could not make my way in the maze. I'll take care of it and make a PR, because I can and because I want.

I agree with renaming type to hint. Do you have a preference for `hint_mappings` or `hint_overrides` ? I think overrides may represent more what we do: no "mapping" is required to the task (there is no mapping by default). Instead, we override hints with other hints.

It's your call.

**Comment by leycec** [comment]
**Wondrousness!** Thanks so much for all you've already done. Please take your merry time. @beartype ain't goin' nowhere – and you've already done more than even five @leycec clones could accomplish. I still want five @leycec clones, though.

> I tried to find myself but I could not make my way in the maze.

:rofl: :sob: 

Indeed, @beartype is the Labyrinth of Pain. I once believed that I could tame the madness and refactor @beartype into sanity. Now, I simply accept I can do nothing and play games about magical superheroes.

> `hint_overrides` ? I think overrides may represent more what we do: no "mapping" is required to the task (there is no mapping by default). Instead, we override hints with other hints.

Yes! So much! "mappings" is, indeed, awkward phrasing here. `hint_transforms` is yet another alternative that might satisfy this use case. That leaves us with:

* `hint_overrides`. **Good.** We override hints with other hints.
* `hint_tranforms`. **Good.** We transform hints into other hints.

I leave this entirely in your capable hands. You care; you are an expert at your craft; and your English is impeccable (whereas my French is trash, because I was born and raised in Los Angeles, California – *where we were not even taught that Canada was a thing*).

Yet again, thank you. You are the Bon Homme that @beartype deserved and finally received. GitHub, you have blessed us!

**Comment by felixchenier** [comment]
Hi @leycec 

I didn't start yet but I'm currently puzzled on a possible recursion bug that may happen with this method. Note that I absolutely don't know how you do the type reduction, I guess you start from a large union of type and decompose it until you get a clean list of types to compare to?

```
my_beartype_decorator = beartype(conf=BeartypeConf(hint_overrides={float: int | float})

@my_beartype_decorator
def function(x: float):
    pass
```

since `float` is replaced by `int | float` in the type checking function, then won't `int | float` be replaced by `int | int | float` then `int | int | int | float` etc.?

Just to be sure that you envisioned this possible recursion problem.

Tx

**Comment by felixchenier** [comment]
@leycec Excellent! Thanks for checking. And sorry that I requested a feature that'll become a bit obsolete (is_pep484_tower) since we can emulate it with hint_overrides={int: numbers.Integral, float: numbers.Real}...

OK I have work to do now!

**Comment by leycec** [comment]
You're most welcome, thanks *so* much for yet another hard-hitting PR, and apologies for not replying sooner. Bugs have been exploding all over the issue tracker like Trudeau in the Canadian polls. `</chuckles>`

Technically, you're absolutely right that `is_pep484_tower` is now obsolete. Pragmatically, I'm still glad we got that. Manual hint overrides *are* more powerful, of course – but they're also more terrifying. This is the price of power. *Unadulterated terror!*

**Comment by leycec** [comment]
**Resolved by 7c5c66d.** Thanks so much for your extreme patience, @felixchenier! 2024 snuck up on us like a cat in the grass you don't see until the last minute, didn't it? But all's well that ends well. I bring good news from the soggy bogs of Ontario.

I'm on the cusp of publicly releasing your great work with @beartype 0.17.0 in a few days. Until then, have an awesome weekend. With any luck, we'll all be shovelling snow and skiing all over the place next week. :snowman_with_snow: :skier:
